lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/randomsearch/RandomSearchStrategy.kt

```diff
@@ -0,0 +1,157 @@
+/*-
+ * #%L
+ * Lincheck
+ * %%
+ * Copyright (C) 2019 JetBrains s.r.o.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.strategy.randomsearch
+
+import org.jetbrains.kotlinx.lincheck.*
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
+import org.jetbrains.kotlinx.lincheck.strategy.ManagedStrategyBase
+import org.jetbrains.kotlinx.lincheck.util.PseudoRandom
+import org.jetbrains.kotlinx.lincheck.verifier.Verifier
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.random.Random
+
+/**
+ * RandomSearchStrategy at first studies codeLocations at the choosen thread
+ * until a blocking event or a thread finish and
+ * then chooses one uniform randomly
+ */
+class RandomSearchStrategy(
```

Comment:
Can we brute force interleavings without repetitions? What does this strategy guarantee?